gcs/error.go

```diff
@@ -1,74 +1,56 @@
-// Copyright (c) 2019 The Decred developers
+// Copyright (c) 2020 The Decred developers
 // Use of this source code is governed by an ISC
 // license that can be found in the LICENSE file.
 
 package gcs
 
-import (
-	"fmt"
-)
-
-// ErrorCode identifies a kind of error.
-type ErrorCode int
+// ErrorKind identifies a kind of error.  It has full support for errors.Is and
+// errors.As, so the caller can directly check against an error kind when
+// determining the reason for an error.
+type ErrorKind string
 
 // These constants are used to identify a specific RuleError.
 const (
 	// ErrNTooBig signifies that the filter can't handle N items.
-	ErrNTooBig ErrorCode = iota
+	ErrNTooBig = ErrorKind("ErrNTooBig")
 
 	// ErrPTooBig signifies that the filter can't handle `1/2**P`
 	// collision probability.
-	ErrPTooBig
+	ErrPTooBig = ErrorKind("ErrPTooBig")
 
 	// ErrBTooBig signifies that the provided Golomb coding bin size is larger
 	// than the maximum allowed value.
-	ErrBTooBig
+	ErrBTooBig = ErrorKind("ErrBTooBig")
 
 	// ErrMisserialized signifies a filter was misserialized and is missing the
 	// N and/or P parameters of a serialized filter.
-	ErrMisserialized
-
-	// numErrorCodes is the maximum error code number used in tests.
-	numErrorCodes
+	ErrMisserialized = ErrorKind("ErrMisserialized")
 )
 
-// Map of ErrorCode values back to their constant names for pretty printing.
-var errorCodeStrings = map[ErrorCode]string{
-	ErrNTooBig:       "ErrNTooBig",
-	ErrPTooBig:       "ErrPTooBig",
-	ErrBTooBig:       "ErrBTooBig",
-	ErrMisserialized: "ErrMisserialized",
-}
-
-// String returns the ErrorCode as a human-readable name.
-func (e ErrorCode) String() string {
-	if s := errorCodeStrings[e]; s != "" {
-		return s
-	}
-	return fmt.Sprintf("Unknown ErrorCode (%d)", int(e))
+// Error satisfies the error interface and prints human-readable errors.
+func (e ErrorKind) Error() string {
+	return string(e)
 }
 
-// Error identifies a filter-related error.  The caller can use type assertions
-// to access the ErrorCode field to ascertain the specific reason for the
-// failure.
+// Error identifies an error related to gcs filters. It has
+// full support for errors.Is and errors.As, so the caller can ascertain the
+// specific reason for the error by checking the underlying error.
 type Error struct {
-	ErrorCode   ErrorCode // Describes the kind of error
-	Description string    // Human readable description of the issue
+	Description string
+	Err         error
```

Comment:
The `Err` field comes first pretty much everywhere else.  Should be the same for consistency.